### PR TITLE
[API][Behat] Add note about behat request changes

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                   |
| Bug fix?        | no                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                      |
| Deprecations?   | no  |
| Related tickets | mentioned in [#Z](https://github.com/Sylius/Sylius/pull/13896)                      |
| License         | MIT                                                          |

### CODE REVIEW

File: UPGRADE-API-1.12.md [hunk 1]
```diff
@@ -225,3 +225,209 @@ The list of changed contexts:
...
+ -    private function request(RequestInterface $request): Response
+ +    public function request(RequestInterface $request): Response
+ ```
+
+#### Defining Section and ContentType changed  
+
+Some of the private methods were removed from `Sylius/Behat/Client/RequestInterface`:
+
+ ```diff
+ -    private static function prepareSection(?string $section): string
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]:
```suggestion
 -   private static function prepareSection(?string $section): string
```

File: UPGRADE-API-1.12.md [hunk 2]
```diff
@@ -225,3 +225,209 @@ The list of changed contexts:
...
+     public function index(string $resource): Response
+    {
+ -      $this->request = Request::index($this->section, $resource, $this->authorizationHeader, $this->getToken());
+ +      $this->request = $this->requestFactory->index($this->section, $resource, $this->authorizationHeader, $this->getToken());
+
+        return $this->request($this->request);
+    }
+ ```
+
+The `public function buildUploadRequest(string $resource): void;` method has been removed, as it's replaced by methods in  `Sylius/Behat/Client/RequestBuilder`.
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]:
```suggestion
The `public function buildUploadRequest(string $resource): void;` method has been removed, as it's replaced by methods in `Sylius/Behat/Client/RequestBuilder`.
```

File: UPGRADE-API-1.12.md [hunk 3]
```diff
@@ -225,3 +225,209 @@ The list of changed contexts:
...
+ -           sprintf('/api/v2/%s%s', self::prepareSection($section), $resource),
+ -           HttpRequest::METHOD_GET,
+ -           $headers
+ -       );
+        
+ +       $builder = RequestBuilder::create(
+ +           sprintf('/api/v2/%s/%s', $section, $resource),
+ +           HttpRequest::METHOD_GET,
+ +       );
+ ```
```
Comment (GSadee) [inline UPGRADE-API-1.12.md]: To simplify the note, I would be for removing this information as it is about private methods, that shouldn't change anything in the end application:
```suggestion
```

File: UPGRADE-API-1.12.md [hunk 4]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
...
+ -       ?string $section,
+ -       string $resource,
+ -       string $authorizationHeader,
+ -       ?string $token = null
+ -   ): self;
+
+ -   public static function custom(string $url, string $method, array $additionalHeaders = [], ?string $token = null): self;
+ ```
+
+All of them have been moved to new `Sylius/Behat/Client/RequestFactory` class.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md]: It's not literally true as these methods are no more static.
```suggestion
All of them have their corresponding implementation in `Sylius/Behat/Client/RequestFactory`.
```
or something like that and I would be for moving this sentence right behind:
>  We removed some methods from `Sylius/Behat/Client/RequestInterface`.

File: UPGRADE-API-1.12.md [hunk 5]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
...
+     public function index(string $resource): Response
+    {
+ -      $this->request = Request::index($this->section, $resource, $this->authorizationHeader, $this->getToken());
+ +      $this->request = $this->requestFactory->index($this->section, $resource, $this->authorizationHeader, $this->getToken());
+
+        return $this->request($this->request);
+    }
+ ```
+
+The `public function buildUploadRequest(string $resource): void;` method has been removed, as it's replaced by methods in `Sylius/Behat/Client/RequestBuilder`.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md]:
```suggestion
The `\Sylius\Behat\Client\ApiClientInterface::buildUploadRequest` method has been removed, as it's replaced by methods in `Sylius/Behat/Client/RequestBuilder`.
```
WDYT? Then there won't be any doubts about where the method was placed.

File: UPGRADE-API-1.12.md [hunk 6]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
 - Sylius\Behat\Context\Api\Shop\PromotionContext
 - Sylius\Behat\Context\Api\Shop\RegistrationContext
 - Sylius\Behat\Context\Api\Shop\ShipmentContext
+
+### Creating Request refactored
+
+We removed some methods from `Sylius/Behat/Client/RequestInterface`.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md]:
```suggestion
We removed some methods from `Sylius\Behat\Client\RequestInterface`.
```
just to be consistent with the entire MD file and you can do the same with the rest of the references.

File: UPGRADE-API-1.12.md [hunk 7]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
...
+ +       $builder->withHeader('HTTP_Authorization', 'Bearer ' . $this->sharedStorage->get('token'));
+ +       $builder->withParameter('owner', $this->iriConverter->getIriFromItem($administrator));
+ +       $builder->withFile('file', new UploadedFile($this->minkParameters['files_path'] . $avatar, basename($avatar)));
+
+ +       $response = $this->client->request($builder->build());
+ ```
+
+As you can see the builder contains the methods that makes the request responsive for additional headers and parameters.
+
+We also changed the `request(RequestInterface $request): Response` method visibility to public.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md]:
```suggestion
We also changed the `\Sylius\Behat\Client\ApiClientInterface::request` method visibility to public.
```

File: UPGRADE-API-1.12.md [hunk 8]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
...
+
+ ```diff
+ -    private function request(RequestInterface $request): Response
+ +    public function request(RequestInterface $request): Response
+ ```
+
+#### Defining ContentType changed
+
+`Sylius/Behat/Client/ContentTypeGuide` is an extended alternative for `resolveHttpRequest()` method.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md]: What `resolveHttpRequest()` is? I cannot find it anywhere 😕

File: UPGRADE-API-1.12.md [hunk 9]
```diff
@@ -225,3 +225,176 @@ The list of changed contexts:
...
+
+ ```
+     public function guide(string $method): string
+    {
+        if ($method === HttpRequest::METHOD_PATCH) {
+            return self::PATCH_CONTENT_TYPE;
+        }
+
+        if ($method === HttpRequest::METHOD_PUT) {
+            return self::LINKED_DATA_JSON_CONTENT_TYPE;
```
Comment (coldic3) [inline UPGRADE-API-1.12.md:360]: Not related to this PR but I don't get why we use JSON-LD only for PUT methods 🤔

File: UPGRADE-API-1.12.md (hunk 8 above)
Comment (TheMilek) [reply UPGRADE-API-1.12.md]: My bad, wrong name - should be resolveHttpMethod() instead 😅

**Review by coldic3 — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Kamil! :1st_place_medal:

**Pull request merged**